### PR TITLE
ENH: allow get_dummies to accept dtype argument

### PR DESCRIPTION
**Pull request by Scorpil**
- [x] closes #18330 (there's no issue for this one)
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Update in version 0.19.0 made `get_dummies` return uint8 values instead of floats (#8725). While I agree with the argument that `get_dummies` should output integers __by default__ (to save some memory), in many cases it would be beneficial for user to choose other dtype.

In my case there was serious performance degradation between versions 0.18 and 0.19. After investigation, reason behind it turned out to be the change to `get_dummies` output type. DataFrame with dummy values was used as an argument to np.dot in an optimization function (second argument was matrix of floats). Since there were lots of iterations involved, and on each iteration np.dot was converting all uint8 values to float64, conversion overhead took unreasonably long time. It is possible to work around this issue by converting dummy columns "manually" afterwards, but it adds unnecessary complexity to the code and is clearly less convenient than calling `get_dummies` with `dtype=float`.

Apart from performance considerations, I can imagine `dtype=bool` to be a common use case.

`get_dummies(data, dtype=None)` is allowed and will return uint8 values to match the [DataFrame](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html) interface (where None allows inferring datatype, which is default behavior).

I've extended the test suite to run all the `get_dummies` tests (except for those that don't deal with internal dtypes, like `test_just_na`) twice, once with `uint8` and once with `float64`.

### CODE REVIEW

File: pandas/core/reshape/reshape.py [hunk 1]
```diff
@@ -725,6 +725,8 @@ def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
     drop_first : bool, default False
         Whether to get k-1 dummies out of k categorical levels by removing the
         first level.
+    dtype : dtype, default np.uint8
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: add a versionadded tag

File: pandas/tests/reshape/test_reshape.py [hunk 2]
```diff
@@ -217,34 +217,36 @@ def test_multiindex(self):
 
 
 class TestGetDummies(object):
```
Comment (jreback) [inline pandas/tests/reshape/test_reshape.py]: instead of doing this define a fixture that returns the various dtypes that you are testing

File: doc/source/whatsnew/v0.22.0.txt [hunk 3]
```diff
@@ -140,7 +140,7 @@ Sparse
 Reshaping
 ^^^^^^^^^
 
--
+- :func:`get_dummies` now supports ``dtype`` argument
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: add a little more expl, add the PR number as the issue number. Move to Other Enhancements section.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by Scorpil** [comment]
All done. Also updated sparse tests to use fixtures as well. And added one test to verify effective dtype is uint8 when dtype argument is None.

File: doc/source/whatsnew/v0.22.0.txt [hunk 4]
```diff
@@ -24,7 +24,17 @@ Other Enhancements
 
 - Better support for :func:`Dataframe.style.to_excel` output with the ``xlsxwriter`` engine. (:issue:`16149`)
 - :func:`pandas.tseries.frequencies.to_offset` now accepts leading '+' signs e.g. '+1h'. (:issue:`18171`)
--
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: make a separate sub-section for this

File: doc/source/whatsnew/v0.22.0.txt [hunk 5]
```diff
@@ -24,7 +24,17 @@ Other Enhancements
...
+- :func:`pandas.get_dummies` now supports ``dtype`` argument, which forces specific dtype for new columns. (:issue:`18330`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: say default is the same (uint8)

File: doc/source/whatsnew/v0.22.0.txt [hunk 6]
```diff
@@ -24,7 +24,17 @@ Other Enhancements
...
+
+.. code-block:: ipython
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: use an ipython block, show the original as well (first)

File: pandas/tests/reshape/test_reshape.py [hunk 7]
```diff
@@ -218,36 +218,46 @@ def test_multiindex(self):
...
+    def df(self, dtype):
+        return DataFrame({'A': ['a', 'b', 'a'],
+                          'B': ['b', 'b', 'c'],
+                          'C': [1, 2, 3]}, dtype=dtype)
 
-    def setup_method(self, method):
-        self.df = DataFrame({'A': ['a', 'b', 'a'],
-                             'B': ['b', 'b', 'c'],
-                             'C': [1, 2, 3]})
+    @pytest.fixture(params=['uint8', 'float64'])
```
Comment (jreback) [inline pandas/tests/reshape/test_reshape.py]: cycle thru more dtypes here that are valid (doesn't have to be all, but include int64, bool, IOW valid for both sparse/dense)

File: pandas/tests/reshape/test_reshape.py [hunk 8]
```diff
@@ -218,36 +218,46 @@ def test_multiindex(self):
...
-                                    2: 0},
-                              'c': {0: 0,
-                                    1: 0,
-                                    2: 1}}, dtype=np.uint8)
-        assert_frame_equal(get_dummies(s_list, sparse=self.sparse), expected)
-        assert_frame_equal(get_dummies(s_series, sparse=self.sparse), expected)
+        expected = DataFrame({'a': [1, 0, 0],
+                              'b': [0, 1, 0],
+                              'c': [0, 0, 1]}, dtype=dtype)
+        result = get_dummies(s_list, **kwargs)
```
Comment (jreback) [inline pandas/tests/reshape/test_reshape.py]: I wouldn't construct the kwargs, actually just pass directly

File: pandas/tests/reshape/test_reshape.py (hunk 7 above)
Comment (jreback) [reply pandas/tests/reshape/test_reshape.py]: add None as well

File: pandas/tests/reshape/test_reshape.py [hunk 9]
```diff
@@ -298,315 +324,314 @@ def test_just_na(self):
...
-        res = get_dummies(s, sparse=self.sparse)
+        res = get_dummies(s, **kwargs)
         exp = DataFrame({'a': {0: 1, 1: 0, 2: 0},
-                         'b': {0: 0, 1: 1, 2: 0}}, dtype=np.uint8)
+                         'b': {0: 0, 1: 1, 2: 0}}, dtype=dtype)
         assert_frame_equal(res, exp)
 
         # Sparse dataframes do not allow nan labelled columns, see #GH8822
-        res_na = get_dummies(s, dummy_na=True, sparse=self.sparse)
+        res_na = get_dummies(s, dummy_na=True, **kwargs)
```
Comment (jreback) [inline pandas/tests/reshape/test_reshape.py]: see my comment above, don't use kwargs generally for passing args to test functions, rather pass directly

File: pandas/tests/reshape/test_reshape.py (hunk 7 above)
Comment (jreback) [reply pandas/tests/reshape/test_reshape.py]: we should prob raise on ``object`` dtype I think.

**Review by jreback — CHANGES_REQUESTED** [review]
looks good generally. thanks for parametrizing the tests!

**Comment by jreback** [comment]
@TomAugspurger

File: doc/source/whatsnew/v0.22.0.txt [hunk 10]
```diff
@@ -24,7 +24,27 @@ Other Enhancements
 
 - Better support for :func:`Dataframe.style.to_excel` output with the ``xlsxwriter`` engine. (:issue:`16149`)
 - :func:`pandas.tseries.frequencies.to_offset` now accepts leading '+' signs e.g. '+1h'. (:issue:`18171`)
--
+
+``get_dummies`` now supports ``dtype`` argument
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: add a ref here as well.

File: doc/source/whatsnew/v0.22.0.txt [hunk 11]
```diff
@@ -24,7 +24,27 @@ Other Enhancements
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:func:`get_dummies` function now accepts ``dtype`` argument, which forces specific dtype for new columns. When ``dtype`` is not specified or equals to ``None``, new columns will have dtype ``uint8`` (as before), so this change is backwards compatible. (:issue:`18330`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: The `` :func:`get_dummies` function now accepts a ``dtype`` argument, which forces a specific dtype for the new columns. The default is ``uint8`` if ``dtype`` is not specified or ``None``.

File: pandas/core/reshape/reshape.py [hunk 12]
```diff
@@ -783,13 +788,25 @@ def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
...
     --------
     Series.str.get_dummies
     """
     from pandas.core.reshape.concat import concat
     from itertools import cycle
 
+    if dtype is None:
+        dtype = np.uint8
+
+    if np.dtype(dtype) is np.dtype('O'):
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: use ``is_object_dtype``

File: pandas/core/reshape/reshape.py [hunk 13]
```diff
@@ -783,13 +788,25 @@ def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
...
+        raise TypeError("'object' is not a valid type for get_dummies")
+
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: this could be a ``ValueError``; also so ``dtype=object is not a valid dtype for get_dummies``

File: pandas/core/reshape/reshape.py [hunk 14]
```diff
@@ -835,17 +852,19 @@ def check_len(item, name):
...
-                                 sparse=sparse, drop_first=drop_first)
+                                 sparse=sparse,
+                                 drop_first=drop_first,
+                                 dtype=dtype)
     return result
 
 
 def _get_dummies_1d(data, prefix, prefix_sep='_', dummy_na=False,
-                    sparse=False, drop_first=False):
+                    sparse=False, drop_first=False, dtype=np.uint8):
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: should this be passthru?

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.22.0.txt [hunk 15]
```diff
@@ -27,6 +27,27 @@ Other Enhancements
 - :class:`pandas.io.formats.style.Styler` now has method ``hide_index()`` to determine whether the index will be rendered in ouptut (:issue:`14194`)
 - :class:`pandas.io.formats.style.Styler` now has method ``hide_columns()`` to determine whether columns will be hidden in output (:issue:`14194`)
 
+``get_dummies`` now supports ``dtype`` argument
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:func:`get_dummies` function now accepts ``dtype`` argument, which forces specific dtype for new columns. When ``dtype`` is not specified or equals to ``None``, new columns will have dtype ``uint8`` (as before), so this change is backwards compatible. (:issue:`18330`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.22.0.txt]: - Can remove `function` after `get_dummies`.
- now accepts *a* dtype argument
- replace *forces* with *specifies a*

Replace the second sentence with

    When ``dtype`` is not specified, the dtype will be ``uint8`` as before.

File: doc/source/whatsnew/v0.22.0.txt [hunk 16]
```diff
@@ -27,6 +27,27 @@ Other Enhancements
...
+
+**Previous behavior**:
+
+.. ipython:: python
+
+   df = pd.DataFrame({'a': [1, 2], 'b': [3, 4], 'c': [5, 6]})
+   pd.get_dummies(df, columns=['c'])
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.22.0.txt]: I think you can remove the "Previous behavior" section since this is backwards compatible.

I'd just do

    pd.get_dummies(df, columns=['c']).dtypes
    pd.get_dummies(df, columns=['c'], dtype=bool).dtypes

File: pandas/core/reshape/reshape.py [hunk 17]
```diff
@@ -697,7 +697,7 @@ def _convert_level_number(level_num, columns):
 
 
 def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
-                columns=None, sparse=False, drop_first=False):
+                columns=None, sparse=False, drop_first=False, dtype=None):
```
Comment (TomAugspurger) [inline pandas/core/reshape/reshape.py]: Any reason not to use `'uint8'` or `np.uint8` here?

File: pandas/core/reshape/reshape.py [hunk 18]
```diff
@@ -728,6 +728,11 @@ def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
 
         .. versionadded:: 0.18.0
 
+    dtype : dtype, default np.uint8
```
Comment (TomAugspurger) [inline pandas/core/reshape/reshape.py]: Can we also accept arguments to `np.dtype` like the string `'i8'`, and handle those appropriately?

File: pandas/tests/reshape/test_reshape.py [hunk 19]
```diff
@@ -218,35 +218,51 @@ def test_multiindex(self):
...
+    def dtype(self, request):
+        return np.dtype(request.param)
+
+    @pytest.fixture(params=['dense', 'sparse'])
+    def sparse(self, request):
+        # params are strings to simplify reading test results,
+        # e.g. TestGetDummies::test_basic[uint8-sparse] instead of [uint8-True]
+        return request.param == 'sparse'
+
+    def effective_dtype(self, dtype):
```
Comment (TomAugspurger) [inline pandas/tests/reshape/test_reshape.py]: If we make the default `np.uint8` you can remove this.

File: pandas/tests/reshape/test_reshape.py [hunk 20]
```diff
@@ -218,35 +218,51 @@ def test_multiindex(self):
...
-
-    def test_basic(self):
+    @pytest.fixture
+    def df(self):
+        return DataFrame({'A': ['a', 'b', 'a'],
+                          'B': ['b', 'b', 'c'],
+                          'C': [1, 2, 3]})
+
+    @pytest.fixture(params=['uint8', 'int64', np.float64, bool, None])
+    def dtype(self, request):
```
Comment (TomAugspurger) [inline pandas/tests/reshape/test_reshape.py]: If we use this fixture in many places it's going to add a ton of tests.

File: pandas/tests/reshape/test_reshape.py [hunk 21]
```diff
@@ -298,315 +319,317 @@ def test_just_na(self):
...
+        res = get_dummies(s, prefix='letter', sparse=sparse, dtype=dtype)
+        exp = DataFrame({'letter_e': [1, 0, 0],
+                         u('letter_%s') % eacute: [0, 1, 1]},
+                        dtype=self.effective_dtype(dtype))
         assert_frame_equal(res, exp)
 
-    def test_dataframe_dummies_all_obj(self):
-        df = self.df[['A', 'B']]
-        result = get_dummies(df, sparse=self.sparse)
+    def test_dataframe_dummies_all_obj(self, df, sparse, dtype):
```
Comment (TomAugspurger) [inline pandas/tests/reshape/test_reshape.py]: What's the benefit of parametrizing by dtype here?

File: pandas/tests/reshape/test_reshape.py [hunk 22]
```diff
@@ -298,315 +319,317 @@ def test_just_na(self):
...
-        expected[cols] = expected[cols].astype(np.uint8)
-        expected = expected[['C', 'from_A_a', 'from_A_b', 'from_B_b',
-                             'from_B_c']]
+        expected[cols] = expected[cols].astype(dtype)
+        expected = expected[['C', 'from_A_a', 'from_A_b',
+                             'from_B_b', 'from_B_c']]
         assert_frame_equal(result, expected)
 
-    def test_dataframe_dummies_prefix_str(self):
+    def test_dataframe_dummies_prefix_str(self, df, sparse, dtype):
```
Comment (TomAugspurger) [inline pandas/tests/reshape/test_reshape.py]: Same question. Is there any relationship between the prefix and dtype? The seem orthogonal.

File: pandas/tests/reshape/test_reshape.py [hunk 23]
```diff
@@ -298,315 +319,317 @@ def test_just_na(self):
...
-                             dtype=np.uint8)
-        expected = expected.astype({"C": np.int64})
+                             columns=['C'] + bad_columns,
+                             dtype=self.effective_dtype(dtype))
+        expected['C'] = [1, 2, 3]
         assert_frame_equal(result, expected)
 
-    def test_dataframe_dummies_subset(self):
-        df = self.df
+    def test_dataframe_dummies_subset(self, df, sparse, dtype):
```
Comment (TomAugspurger) [inline pandas/tests/reshape/test_reshape.py]: Same question: Any interaction between `subset` and `dtype`?

File: pandas/tests/reshape/test_reshape.py [hunk 24]
```diff
@@ -298,315 +319,317 @@ def test_just_na(self):
...
-        cols = ['from_A_a', 'from_A_b']
-        expected[cols] = expected[cols].astype(np.uint8)
+        columns = ['from_A_a', 'from_A_b']
+        expected[columns] = expected[columns].astype(dtype)
         assert_frame_equal(result, expected)
 
-    def test_dataframe_dummies_prefix_sep(self):
-        df = self.df
-        result = get_dummies(df, prefix_sep='..', sparse=self.sparse)
+    def test_dataframe_dummies_prefix_sep(self, df, sparse, dtype):
```
Comment (TomAugspurger) [inline pandas/tests/reshape/test_reshape.py]: Same question :)

**Review by TomAugspurger — COMMENTED** [review]
Thanks.

I'll finish taking a look later but my only real concern is how many times you parametrize by the dtype. It's great to do that for some tests like `test_basic_dtype` and a few others, but I'm not sure about all the prefix / sep tests. Do you have specific concerns that you're trying to test there?

File: pandas/core/reshape/reshape.py (hunk 17 above)
Comment (Scorpil) [reply pandas/core/reshape/reshape.py]: I've tried to mirror API of DataFrame, Series, Panel etc. where passing None explicitly is allowed and means "dtype will be inferred".

File: pandas/tests/reshape/test_reshape.py (hunk 20 above)
Comment (jreback) [reply pandas/tests/reshape/test_reshape.py]: these are all quick, so this is ok

File: pandas/core/reshape/reshape.py (hunk 18 above)
Comment (jreback) [reply pandas/core/reshape/reshape.py]: he is using ``np.dtype()``?

Comment (Scorpil) [reply pandas/core/reshape/reshape.py]: Yes, that should work already, I'll add it to the tests.

File: pandas/tests/reshape/test_reshape.py (hunk 20 above)
Comment (Scorpil) [reply pandas/tests/reshape/test_reshape.py]: Initially I only had 'uint8' and 'float64', but @jreback reasonably suggested to add some more. What would be a good balance here? If I'll remove usage of this fixture from all the unrelated tests like prefix / separator tests, and move None to separate stand-alone test, would ['uint8', 'i8', np.float64, bool] be OK? Still x4 number of tests, but each item uses a different way to specify dtype, so i think it's meaningful set of fixtures.

Comment (jreback) [reply pandas/tests/reshape/test_reshape.py]: i think toms point is to not apply this fixture to every test (just relevant ones)

File: pandas/core/reshape/reshape.py (hunk 17 above)
Comment (Scorpil) [reply pandas/core/reshape/reshape.py]: @jreback @TomAugspurger So this is the last question to answer. Do you accept my argument about None or should I change it to np.uint8?

File: doc/source/whatsnew/v0.22.0.txt [hunk 25]
```diff
@@ -27,6 +27,17 @@ Other Enhancements
 - :class:`pandas.io.formats.style.Styler` now has method ``hide_index()`` to determine whether the index will be rendered in ouptut (:issue:`14194`)
 - :class:`pandas.io.formats.style.Styler` now has method ``hide_columns()`` to determine whether columns will be hidden in output (:issue:`14194`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: can you add a ref here

File: pandas/core/reshape/reshape.py (hunk 17 above)
Comment (jreback) [reply pandas/core/reshape/reshape.py]: this is ok here, it follows a similar style elsewhere

File: pandas/core/reshape/reshape.py [hunk 26]
```diff
@@ -783,13 +788,25 @@ def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
...
     See Also
     --------
     Series.str.get_dummies
     """
     from pandas.core.reshape.concat import concat
     from itertools import cycle
 
+    if dtype is None:
+        dtype = np.uint8
+
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: need a ``dtype = np.dtype(dtype)``

File: pandas/core/reshape/reshape.py (hunk 14 above)
Comment (jreback) [reply pandas/core/reshape/reshape.py]: ?

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.22.0.txt (hunk 25 above)
Comment (Scorpil) [reply doc/source/whatsnew/v0.22.0.txt]: I'm sorry, not used to work with sphinx. Do you mean something like this here:

```
- :func:`get_dummies` now supports ``dtype`` argument, see :ref:`here <whatsnew_0220.enhancements.get_dummies_dtype>` for more  (:issue: `18330`)
```

and then this before the actual description block:
```
.. _whatsnew_0220.enhancements.get_dummies_dtype
```
?

File: pandas/core/reshape/reshape.py (hunk 14 above)
Comment (Scorpil) [reply pandas/core/reshape/reshape.py]: Right, missed this one. Idea was to treat this one as internal and allow "wrapper" to set dtype, but passthru has it's advantages and I don't mind ether way, so I'll move dtype-related conversions to this method.

File: pandas/tests/reshape/test_reshape.py [hunk 27]
```diff
@@ -298,270 +319,258 @@ def test_just_na(self):
...
+        expected = expected[['C', 'from_A_a', 'from_A_b',
+                             'from_B_b', 'from_B_c']]
         assert_frame_equal(result, expected)
 
-    def test_dataframe_dummies_prefix_str(self):
+    def test_dataframe_dummies_prefix_str(self, df, sparse):
         # not that you should do this...
-        df = self.df
-        result = get_dummies(df, prefix='bad', sparse=self.sparse)
+        df[['C']] = df[['C']].astype(np.uint8)
```
Comment (TomAugspurger) [inline pandas/tests/reshape/test_reshape.py]: Why the change here?

File: pandas/tests/reshape/test_reshape.py [hunk 28]
```diff
@@ -298,270 +319,258 @@ def test_just_na(self):
...
+        columns = ['from_A_a', 'from_A_b', 'from_B_b', 'from_B_c']
+        expected[columns] = expected[columns].astype(np.uint8)
         assert_frame_equal(result, expected)
 
-    def test_dataframe_dummies_with_na(self):
-        df = self.df
+    def test_dataframe_dummies_with_na(self, df, sparse, dtype):
         df.loc[3, :] = [np.nan, np.nan, np.nan]
-        result = get_dummies(df, dummy_na=True, sparse=self.sparse)
+        result = get_dummies(df, dummy_na=True,
```
Comment (TomAugspurger) [inline pandas/tests/reshape/test_reshape.py]: Did you add the sorting because the output changed, or to make the test easier to write?

I slightly prefer the explicit ordering rather than sorting, though that'll be covered elsewhere so changing it isn't a huge deal.

**Review by TomAugspurger — APPROVED** [review]
Gave another quick glance, and things look good here.

File: pandas/tests/reshape/test_reshape.py (hunk 27 above)
Comment (Scorpil) [reply pandas/tests/reshape/test_reshape.py]: This test is a bit weird... Having 2 columns with identical names caused ValueError when I tried `expected['C'] = expected['C']`. Now, when you mentioned it, I see in diff that `expected = expected.astype({"C": np.int64})` should work, I'll put it back.

File: pandas/tests/reshape/test_reshape.py (hunk 28 above)
Comment (Scorpil) [reply pandas/tests/reshape/test_reshape.py]: Just a bit easier to write.

File: doc/source/whatsnew/v0.22.0.txt [hunk 29]
```diff
@@ -28,6 +28,20 @@ Other Enhancements
...
 - Improved wording of ``ValueError`` raised in :func:`to_datetime` when ``unit=`` is passed with a non-convertible value (:issue:`14350`)
+- :func:`get_dummies` now supports ``dtype`` argument, see :ref:`here <whatsnew_0220.enhancements.get_dummies_dtype>` for more (:issue:`18330`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: you can remove this line, its already covered in the sub-section. move the sub-section *before* other enhancements

File: pandas/tests/reshape/test_reshape.py [hunk 30]
```diff
@@ -218,35 +218,51 @@ def test_multiindex(self):
...
+        # params are strings to simplify reading test results,
+        # e.g. TestGetDummies::test_basic[uint8-sparse] instead of [uint8-True]
+        return request.param == 'sparse'
+
+    def effective_dtype(self, dtype):
+        if dtype is None:
+            return np.uint8
+        return dtype
+
+    def test_throws_on_dtype_object(self, df):
```
Comment (jreback) [inline pandas/tests/reshape/test_reshape.py]: throws -> raises

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. small doc changes. have a look in reshaping.rst if any doc updates are needed.

File: doc/source/reshaping.rst [hunk 31]
```diff
@@ -634,6 +634,17 @@ When a column contains only one level, it will be omitted in the result.
...
 
+By default new columns will have ``np.uint8`` dtype. To choose another dtype use ``dtype`` argument:
+
+.. ipython:: python
+
+    df = pd.DataFrame({'A': list('abc'), 'B': [1.1, 2.2, 3.3]})
+
+    pd.get_dummies(df, dtype=bool).dtypes
+
+.. versionadded:: 0.22.0
```
Comment (jreback) [inline doc/source/reshaping.rst]: ca you move to before the example

File: doc/source/whatsnew/v0.22.0.txt [hunk 32]
```diff
@@ -17,6 +17,21 @@ New features
 -
 -
 
+
+.. _whatsnew_0210.enhancements.get_dummies_dtype:
+
+``get_dummies`` now supports ``dtype`` argument
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The :func:`get_dummies` now accepts a ``dtype`` argument, which specifies a specific dtype for the new columns. When ``dtype`` is not specified or ``None``, the dtype will be ``uint8`` as before. (:issue:`18330`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: just say the default remains ``uint8``

File: doc/source/reshaping.rst [hunk 33]
```diff
@@ -240,7 +240,7 @@ values will be set to ``NaN``.
    df3
    df3.unstack()
 
-.. versionadded: 0.18.0
+.. versionadded:: 0.18.0
```
Comment (Scorpil) [inline doc/source/reshaping.rst]: This was a typo, right? There are couple more places where second double column is missing:
```
pandas/core/frame.py
4516:            .. versionadded: 0.18.0
4679:            .. versionadded: 0.16.1

pandas/core/generic.py
968:            .. versionadded: 0.21.0

pandas/core/series.py
1629:            .. versionadded: 0.19.0
2216:            .. versionadded: 0.18.0

pandas/core/tools/datetimes.py
117:        .. versionadded: 0.18.1
143:        .. versionadded: 0.16.1
181:        .. versionadded: 0.20.0
187:        .. versionadded: 0.22.0

pandas/tseries/offsets.py
778:    .. versionadded: 0.16.1
882:    .. versionadded: 0.18.1
```

File: pandas/core/reshape/reshape.py [hunk 34]
```diff
@@ -835,20 +846,30 @@ def check_len(item, name):
...
 
 def _get_dummies_1d(data, prefix, prefix_sep='_', dummy_na=False,
-                    sparse=False, drop_first=False):
+                    sparse=False, drop_first=False, dtype=None):
     # Series avoids inconsistent NaN handling
     codes, levels = _factorize_from_iterable(Series(data))
 
+    if dtype is None:
+        dtype = np.uint8
+    else:
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]:
```
if dtype is None
    dtype = np.uint8
dtype = np.dtype(dtype)
```

a bit more idiomatic

**Review by jreback — CHANGES_REQUESTED** [review]
small comments. ping on green.

File: doc/source/reshaping.rst (hunk 33 above)
Comment (jreback) [reply doc/source/reshaping.rst]: hmm yes looks that way. would be great if you can update those! (if you *really* want to could also add a lint rule to search for these and fail the build if they are found) (also in doc dir too).

Comment (Scorpil) [reply doc/source/reshaping.rst]: Separate PR or this will do?

Comment (jreback) [reply doc/source/reshaping.rst]: separate PR prob better. (the one you changed already is fine). I think we DO want to add some more generic checks for these formatting tags, I guess sphinx doesn't complain

Comment (Scorpil) [reply doc/source/reshaping.rst]: Yeah, it's just comments for sphinx. I'll create an issue then, and see what I can do when I have time to look into it. Or somebody will pick it up before that, which is also fine :)

Comment (Scorpil) [reply doc/source/reshaping.rst]: #18425

**Comment by Scorpil** [comment]
@jreback it's green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Scorpil nice patch! keem em coming!
